Initial frontend work

strburst opened the pull request:
<!--
It's ok if the pull isn't ready to be merged yet; just prepend "[WIP]" to the title. This is a good idea if you want some preliminary feedback.

You can always insert screenshots if they will be helpful.

If this pull doesn't consist solely of superficial changes, make sure that all the questions below are answered (even with N/A) and that all the todos are checked.
-->

## Changes <!-- What does this pull request do/change? -->
This pull does a few things:
- [x] Switches the templating engine to haml
- [x] Gets npm-pipeline-rails installed and working
- [x] Customize gulp setup with at least babel and eslint

I was originally going to also attempt to replicate the functionality of the prototype, but I think that would be too much for one pull.

## Context <!-- Is there some background that reviewers need? -->
- Quick demonstration of haml: http://haml.info/tutorial.html
- Haml reference page: http://haml.info/docs/yardoc/file.REFERENCE.html
- Prototype: https://jdteamacetabulum.github.io/Active-Learning/index.html
- npm-pipeline-rails: https://github.com/rstacruz/npm-pipeline-rails

## Testing <!-- What's the best way to try out the changes manually? -->
Create a new controller and see that the example CSS and Javascript in the gulp folder is being injected properly.

## Previous discussions <!-- Are there any relevant issues (user stories, bug reports, questions, etc.)? Link them here. -->
N/A.

## Pre-merge checklist <!-- Things that should be done before merging. Feel free to add more boxes, especially if this is a WIP. -->

- [x] Build succeeds locally and in CI
- [ ] Proper documentation
- [ ] Tests <!-- if unnecessary, check and say N/A -->
- [x] Git history cleaned up <!-- Mistake commits and unnecessary merge commits rebased away, proper commit message conventions followed: http://chris.beams.io/posts/git-commit/ -->

strburst committed:
Make minor tweaks to Gemfile organization

This commit also includes some Gemfile.lock changes that are reflected
on Linux.

strburst committed:
Change the templating engine to haml

strburst committed:
Modify shebangs to work on a Linux system

strburst committed:
Add npm-pipeline-rails and auto-generated files

This commit is what `./bin/rails generate npm_pipeline:gulp` from
https://github.com/rstacruz/npm-pipeline-rails/blob/master/docs/gulp.md
spits out mostly unmodified.

strburst committed:
Remove coffeescript and revert to vanilla js

This may cause the app to crash if there are cached coffeescript assets.
Remove these with `rake tmp:cache:clear`.

strburst committed:
Fix filenames so that example assets are found

strburst committed:
Set up eslint through gulp

strburst committed:
Add sass support to gulpfile

strburst committed:
Add a simple clean task to gulp

strburst committed:
Also ignore npm debug logs and node repl history

strburst committed:
Polish the example app.scss and app.js files

strburst committed:
Add babel and uglify to javascript build pipeline

fvosburgh commented on Gemfile:
```diff
@@ -30,9 +32,16 @@ gem 'jbuilder', '~> 2.5'
+# Use npm/gulp to do the frontend build
+gem 'npm-pipeline-rails', '~> 1.7'
+# Lightweight css boilerplate
+gem 'skeleton-rails', :git => 'https://github.com/helios-technologies/skeleton-rails'
```
If we're going to use `skeleton`, then we need to remove `sass-rails`.

strburst replied:
Assuming you mean `npm-pipeline-rails`; I just pushed a fix.

fvosburgh commented on Gemfile.lock:
```diff
@@ -79,8 +87,14 @@ GEM
     minitest (5.10.1)
     multi_json (1.12.1)
     nio4r (1.2.1)
+    nokogiri (1.7.0.1)
```
Why do we have two versions of `nokogiri`?

fvosburgh replied:
Nvm, I figured this one out. There's different versions of nokogiri and other gems because @rmkeezer is difficult and wants to develop on Windows lol

fvosburgh commented on Gemfile.lock:
```diff
@@ -110,20 +124,24 @@ GEM
       actionpack (>= 4.0)
       activesupport (>= 4.0)
       sprockets (>= 3.0.0)
+    sqlite3 (1.3.13)
```
We need to get rid of `sqlite` as well. Should be done in separate PR though. I'll do it for #10

fvosburgh reviewed (commented):
Other than a few custodial comments, this looks good to me. I don't think a readme for how the frontend works is entirely necessary if you can provide some reading on the subject. Time is valuable, and writing a development guide might throw our velocity off

strburst commented:
I just realized the only thing you would need to do to make it work is run `npm install` but `npm-pipeline-rails` actually sets it up to do that automatically, so yeah, no additional instructions are actually needed.

This looks ready to merge to me.

strburst committed:
Remove unneeded sass gem

The frontend build through gulp/npm is already set up to compile sass,
so it's not needed to have this functionality again with the ruby gem.
